File: content/problems/1144-decrease-elements-to-make-array-zigzag/solutions/two-parity-valley-cost/solution.cpp
class Solution {
    int cost(vector<int>& nums,int parity){
        int moves=0;
        for(int i=parity;i<(int)nums.size();i+=2){
            int left=i?nums[i-1]:1000000000,right=i+1<(int)nums.size()?nums[i+1]:1000000000;
            moves+=max(0,nums[i]-min(left,right)+1);
        }
        return moves;
    }
    public:int movesToMakeZigzag(vector<int>& nums){
        return min(cost(nums,0),cost(nums,1));
    }
};

File: content/problems/1144-decrease-elements-to-make-array-zigzag/solutions/two-parity-valley-cost/Solution.kt
class Solution {
    fun movesToMakeZigzag(nums:IntArray):Int=minOf(cost(nums,0),cost(nums,1));
    private fun cost(nums:IntArray,parity:Int):Int{
        var moves=0;
        for(i in parity until nums.size step 2){
            val left=if(i>0)nums[i-1] else 1000000000;
            val right=if(i+1<nums.size)nums[i+1] else 1000000000;
            moves+=maxOf(0,nums[i]-minOf(left,right)+1)
        };
        return moves
    }
}
